misc/python/materialize/cli/mzconduct.py

```diff
@@ -946,27 +947,115 @@ def run(self, comp: Composition, workflow: Workflow) -> None:
             raise Failed(f"Unable to kill container {container_id}: {e}")
 
 
-@Steps.register("chaos-delay-docker")
-class ChaosDelayDockerStep(WorkflowStep):
-    """Delay the incoming and outgoing network traffic for a Docker service.
+class ChaosNetemStep(WorkflowStep):
+    """Base class for running netem chaos against a Docker container.
 
-    Params:
-        service: Docker service to delay, will be used to grep for container id
-                 NOTE: service name must be unique to correctly match the container id
-        delay: milliseconds to delay network traffic (default: 100ms)
+    We use pumba (a chaos testing tool for Docker) to run the various netem tests.
+    pumba only supports Docker container names (not container ids), meaning that
+    we have to pass the exact container name to these steps. We should fix this in
+    the future.
     """
 
-    def __init__(self, service: str, delay: int = 100) -> None:
-        self._service = service
-        self._delay = delay
+    def __init__(self, duration: int):
+        self._duration = duration
 
     def run(self, comp: Composition, workflow: Workflow) -> None:
+        if self._duration == -1:
+            # If duration isn't provided, run for 7 days in a non-blocking thread.
+            duration = 10080
+        else:
+            duration = self._duration
+
+        cmd = self.get_cmd(duration).split()
+        netem_thread = Thread(target=self.threaded_netem, args=(cmd,), daemon=True)
+        netem_thread.start()
+
+    def get_cmd(self, duration: int) -> str:
+        pass
+
+    def threaded_netem(self, cmd: List[str]) -> None:
         try:
-            container_id = comp.get_container_id(self._service)
-            cmd = f"docker exec {container_id} tc qdisc add dev eth0 root netem delay {self._delay}ms".split()
             spawn.runv(cmd)
         except subprocess.CalledProcessError as e:
-            raise Failed(f"Unable to delay container {container_id}: {e}")
+            raise Failed(f"Unable to run netem chaos command: {e}")
+
+
+@Steps.register("chaos-delay-docker")
+class ChaosDelayDockerStep(ChaosNetemStep):
+    """Delay the egress network traffic for a Docker service.
+    """
+
+    def __init__(
+        self, container: str, duration: int = -1, delay: int = 250, jitter: int = 250,
+    ) -> None:
+        super().__init__(duration=duration)
+        self._container = container
+        self._duration = duration
```

Comment:
It's I think technically wrong that this is setting duration both in the `super()` call and here.

Reply:
yes! will remove